File: frontend/src/lib/group-repos.ts
import type { Repository } from '@/lib/hooks';

export interface ConnectionGroup {
  connectionId: string;
  provider: string;
  username: string;
  repos: Repository[];
}

/** Group repositories by their git-host connection, groups sorted by provider. */
export function groupByConnection(repos: Repository[]): ConnectionGroup[] {
  const groups = new Map<string, ConnectionGroup>();
  for (const repo of repos) {
    const group =
      groups.get(repo.connectionId) ?? {
        connectionId: repo.connectionId,
        provider: repo.connection.provider,
        username: repo.connection.username,
        repos: [],
      };
    group.repos.push(repo);
    groups.set(repo.connectionId, group);
  }
  return [...groups.values()].sort((a, b) => a.provider.localeCompare(b.provider));
}

File: android/app/src/main/java/space/grigteo/lemniscate/core/GroupRepos.kt
package space.grigteo.lemniscate.core

import space.grigteo.lemniscate.core.api.RepositoryDto

data class ConnectionGroup(
    val connectionId: String,
    val provider: String,
    val username: String,
    val repos: List<RepositoryDto>,
)

/** Group repositories by their git-host connection, groups sorted by provider. */
fun groupByConnection(repos: List<RepositoryDto>): List<ConnectionGroup> {
    val groups = LinkedHashMap<String, MutableList<RepositoryDto>>()
    for (repo in repos) groups.getOrPut(repo.connectionId) { mutableListOf() }.add(repo)
    return repos
        .distinctBy { it.connectionId }
        .map { repo ->
            ConnectionGroup(
                connectionId = repo.connectionId,
                provider = repo.connection.provider,
                username = repo.connection.username,
                repos = groups.getValue(repo.connectionId),
            )
        }
        .sortedBy { it.provider }
}
